### PR TITLE
Update wasmi to 0.14.0

### LINKED ISSUE #1605: Update wasmi to 0.14.0
### Problem to Solve

blocked by #1604

Wasmi 0.14.0 has new API that allows to set/get global within host call. It is also the last version with `virtual_memory` feature required by lazy_pages mechanism.

### Possible Solution

The implementation is similar to the previous wasmtime backend.

### Notes

_No response_

### PR DESCRIPTION
**Pull request by gshep**
Resolves #1605 

@gear-tech/dev

### CODE REVIEW

**Review by shamilsan — COMMENTED** [review]
I am sorry but PR is too huge and seems to have unrelated changes. Need more info in details what've done. 

- What is `GetGasAmount` for?
- Why funcs have been reworked (they seem to become overcomplicated)?
- Why `MemoryWrap` has been expanded?
- Etc. etc.

I'd divide the PR into several more atomic ones.

**Comment by gshep** [comment]
> I am sorry but PR is too huge and seems to have unrelated changes. Need more info in details what've done.
> 
>     * What is `GetGasAmount` for?
> 
>     * Why funcs have been reworked (they seem to become overcomplicated)?
> 
>     * Why `MemoryWrap` has been expanded?
> 
>     * Etc. etc.
> 
> 
> I'd divide the PR into several more atomic ones.

I absolutely agree that the small PR - the better, however splitting this one into several looks almost impossible. The main reason is that wasmi API changed drastically and looks similar to wasmtime's. Nevertheless I tried to made atomic commits as possible - so one can review the PR consecutively commit after commit.

1) `GetGasAmount` is just required if `execute` finishes with an error. Previous `execute` implementation used mutable reference to `Ext` so the caller could use it after the function call. With this implementation requiring moved Ext instance an entity implementing GetGasAmount is returned in a case of an error. For successful execution Ext instance returned in BackendReport.

(2) and (3) both are consequences of changing wasmi-API. Previously we had wasmtime backend so it is better to compare with it: https://github.com/gear-tech/gear/commit/094f8fcd648926791b8dfb95997d70357b6262a1#diff-ad9c212bf362102811a274461a869e0e10208004def3c63367cb36aceb4bbfd3

If there are any other questions please provide them. Or we can just discuss the changes on a call.

File: core-backend/wasmi/src/funcs/mod.rs [hunk 1]
```diff
@@ -0,0 +1,1650 @@
...
+            Err(DummyHostError.into())
+        };
+
+        Func::wrap(store, func)
+    }
+
+    pub fn exit_code(store: &mut Store<HostState<E>>, forbidden: bool) -> Func {
+        let func = move |mut caller: wasmi::Caller<'_, HostState<E>>| {
+            let host_state = host_state_mut!(caller);
+            if forbidden {
```
Comment (grishasobol) [inline core-backend/wasmi/src/funcs/mod.rs:610]: Can we do macros for `forbidden if ` ? I think it's very suitable here

**Review by grishasobol — APPROVED** [review]
Ok for me

File: core-backend/wasmi/src/funcs/mod.rs (hunk 1 above)
Comment (gshep) [reply core-backend/wasmi/src/funcs/mod.rs]: I'm not sure actually but anyway `exit_if` introduced. Please check how it looks now.

Comment (grishasobol) [reply core-backend/wasmi/src/funcs/mod.rs]: Yes look ok)

**Review by ark0f — CHANGES_REQUESTED** [review]

File: core-backend/common/src/lib.rs [hunk 2]
```diff
@@ -124,6 +124,10 @@ pub trait IntoExtInfo {
     fn trap_explanation(&self) -> Option<TrapExplanation>;
 }
 
+pub trait GetGasAmount {
```
[the remark this answers is not in the document]
Comment (gshep) [reply core-backend/common/src/lib.rs:127]: I tried to follow substrate's naming style: there is a trait `Get<T>`. Frankly speaking I could use it but didn't want to add new dependency. So I would live the name untouched.

File: core/src/gas.rs [hunk 3]
```diff
@@ -46,9 +46,9 @@ pub enum ChargeResult {
 
 /// Gas counter with some predefined maximum gas.
 ///
-/// `Clone` and `Copy` traits aren't implemented for the type (however could be)
-/// in order to make the data only moveable, preventing implicit/explicit copying.
-#[derive(Debug)]
+/// `Copy` trait isn't implemented for the type (however could be)
+/// in order to make the data only moveable, preventing implicit copying.
+#[derive(Clone, Debug)]
```
[the remark this answers is not in the document]
Comment (gshep) [reply core/src/gas.rs:51]: To help implementing `GetGasAmount` and get `GasAmount` from `GasCounter` instance.

File: core/src/memory.rs [hunk 4]
```diff
@@ -301,7 +301,7 @@ pub trait Memory {
     fn data_size(&self) -> usize;
 
     /// Returns native addr of wasm memory buffer in wasm executor
-    fn get_buffer_host_addr(&self) -> Option<HostPointer> {
+    fn get_buffer_host_addr(&mut self) -> Option<HostPointer> {
```
[the remark this answers is not in the document]
Comment (gshep) [reply core/src/memory.rs:304]: We call `data_mut` on wasmi's `Memory` instance which requires mutable `self`. Actually this is a moment to discuss I touched on the call.

File: core-backend/common/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (gshep) [reply core-backend/common/src/lib.rs:127]: We could but the trait used for different types so there is no sense to do it.

File: core-backend/wasmi/Cargo.toml [hunk 5]
```diff
@@ -8,10 +8,10 @@ license = "GPL-3.0"
 [dependencies]
 gear-core = { path = "../../core" }
 gear-core-errors = { path = "../../core-errors", features = ["codec"] }
-gear-backend-common = { path = "../common" }
+gear-backend-common = { path = "../common", features = ["mock"] }
```
[the remark this answers is not in the document]
Comment (gshep) [reply core-backend/wasmi/Cargo.toml:11]: Why? `MockExt` is required for tests but it is under the feature if I'm not mistaken.

File: core/src/memory.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ark0f) [reply core/src/memory.rs:304]: I discussed it with @grishasobol and we vote to leave it as in PR

File: core-backend/wasmi/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ark0f) [reply core-backend/wasmi/Cargo.toml:11]: Add `gear-backend-common` to dev deps with the feature then

[the remark this answers is not in the document]
Comment (gshep) [reply core-backend/wasmi/Cargo.toml:11]: The crate is also required as a normal dependency

File: core-backend/wasmi/src/env.rs [hunk 6]
```diff
@@ -19,416 +19,261 @@
...
-
-        let result = (self.defined_host_functions.funcs[index])(self.state, &args);
-        match result {
-            Ok(value) => Ok(match value {
-                ReturnValue::Value(v) => Some(v),
-                ReturnValue::Unit => None,
-            }),
-            Err(_e) => Err(Trap::Host(Box::new(DummyHostError))),
-        }
+impl fmt::Display for Error {
```
[the remark this answers is not in the document]
Comment (gshep) [reply core-backend/wasmi/src/env.rs]: ok

File: core-backend/wasmi/src/env.rs [hunk 7]
```diff
@@ -19,416 +19,261 @@
...
+            .define("env", "memory", memory)
+            .map_err(|e| Error::from((ext.gas_amount(), Linking(e))))?;
+
+        let forbidden_funcs =
+            (!ext.forbidden_funcs().is_empty()).then(|| ext.forbidden_funcs().clone());
+        let functions = funcs_tree::build(&mut store, memory, forbidden_funcs);
+        for (name, function) in functions {
+            linker
+                .define("env", name, function)
+                .map_err(|e| Error::from((ext.gas_amount(), Linking(e))))?;
```
[the remark this answers is not in the document]
Comment (gshep) [reply core-backend/wasmi/src/env.rs]: ok

File: core-backend/wasmi/src/env.rs [hunk 8]
```diff
@@ -19,416 +19,261 @@
...
-            Ok(inst) => inst.not_started_instance().clone(),
-            Err(e) => return Err(ModuleInstantiation(e)),
-        };
+        *store.state_mut() = Some(runtime);
+
+        let (ext, memory_wrap, termination) = {
+            let instance_pre = match linker.instantiate(&mut store, &module) {
+                Ok(i) => i,
+                Err(e) => {
+                    let gas_amount = store
```
[the remark this answers is not in the document]
Comment (gshep) [reply core-backend/wasmi/src/env.rs]: ko

File: core-backend/sandbox/src/env.rs [hunk 9]
```diff
@@ -56,6 +57,30 @@ pub enum SandboxEnvironmentError {
     StackEnd(StackEndError),
 }
 
+#[derive(Debug)]
+pub struct Error {
+    gas_amount: GasAmount,
+    error: SandboxEnvironmentError,
+}
+
+impl fmt::Display for Error {
```
[the remark this answers is not in the document]
Comment (gshep) [reply core-backend/sandbox/src/env.rs]: ok

**Review by shamilsan — APPROVED** [review]

**Review by ukint-vs — APPROVED** [review]

File: core-backend/wasmi/Cargo.toml (hunk 5 above)
[the remark this answers is not in the document]
Comment (ark0f) [reply core-backend/wasmi/Cargo.toml:11]: You can leave it as normal dep without the feature and also add it to dev section with feature

[the remark this answers is not in the document]
Comment (gshep) [reply core-backend/wasmi/Cargo.toml:11]: ok

**Review by ark0f — APPROVED** [review]

**Pull request merged**